File: src/org/heli/ChopperInfo.java
/** Copyright 2015, Dan and Sasha
 * 
 */
package org.heli;

/** The world uses this class to maintain all necessary flight information
 * about a chopper.
 * @author Daniel LaFuze
 *
 */
public class ChopperInfo {
	
	private World myWorld;
	
	int chopperID;
	
	double currentFuel_kg;
	
	private static final double THRUST_PER_RPM = 11.1111; // N (kg * m/s^2)
	
	public static final double MAX_MAIN_ROTOR_SPEED = 400.0; // RPM
	
	private static final double EARTH_ACCELERATION = -9.80665; // m/s^2
	
	public static final double MAX_TAIL_ROTOR_SPEED = 120.0; // RPM
	public static final double STABLE_TAIL_ROTOR_SPEED = 100.0; // RPM
	public static final double MIN_TAIL_ROTOR_SPEED = 80.0; // RPM
	
	public static final double MAX_TILT_MAGNITUDE = 5.0; // Degrees
	
	private static final double MAX_MAIN_ROTOR_DELTA = 60.0; // RPM per Second
	
	private static final double MAX_TAIL_ROTOR_DELTA = 15.0; // RPM per Second
	
	private static final double MAX_TILT_DELTA = 3.0; // Degrees per second
	
	private static final double FUEL_PER_REVOLUTION = 1.0 / 60.0; // Liters
	
	private static final double ROTATION_PER_TAIL_RPM = 3.0; // degrees per second
	
	private double actMainRotorSpeed_RPM;
	
	private double actTailRotorSpeed_RPM;
	
	private double actTilt_Degrees;
	
	private double desMainRotorSpeed_RPM;
	
	private double desTailRotorSpeed_RPM;
	
	private double desTilt_Degrees;
	
	private double remainingFuel_kg;
	
	private boolean takenOff = false;
	
	// In meters per second squared
	private Point3D actAcceleration_ms2;
	
	// In meters per second
	private Point3D actVelocity_ms;
	
	// In meters
	private Point3D actPosition_m;
	
	double heading_Degrees;

	/** Constructs a ChopperInfo object with a chopper on the ground
	 * 
	 * @param myWorld The world (for interface)
	 * @param chopperID To get information about chopper -- such as weight
	 * @param startPos Where will the chopper start
	 * @param startHeading Which way is chopper facing?
	 */
	public ChopperInfo(World world, int id, Point3D startPos, double startHeading) {
		myWorld = world;
		chopperID = id;
		actPosition_m = startPos;
		heading_Degrees = startHeading;
		actMainRotorSpeed_RPM = 0.0;
		actTailRotorSpeed_RPM = 0.0;
		actTilt_Degrees = 0.0;
		desMainRotorSpeed_RPM = 0.0;
		desTailRotorSpeed_RPM = 0.0;
		desTilt_Degrees = 0.0;
		actAcceleration_ms2 = new Point3D(0.0, 0.0, 0.0);
		actVelocity_ms = new Point3D(0.0, 0.0, 0.0);
		StigChopper thisChopper = myWorld.getChopper(chopperID);
		remainingFuel_kg = 0.0;
		if (thisChopper != null)
		{
			remainingFuel_kg = thisChopper.fuelCapacity();
		}
	}
	
	public Point3D getPosition()
	{
		return actPosition_m;
	}
	
	public double getHeading()
	{
		return heading_Degrees;
	}
	
	public double getTilt()
	{
		return actTilt_Degrees;
	}
	
	public void requestMainRotorSpeed(double newSpeed)
	{
		desMainRotorSpeed_RPM = newSpeed;
	}
	
	public void requestTailRotorSpeed(double newSpeed)
	{
		desTailRotorSpeed_RPM = newSpeed;
	}
	
	public void requestTiltLevel(double newTilt)
	{
		desTilt_Degrees = newTilt;
	}
	
	private void updateMainRotorSpeed(double elapsedTime)
	{
		double deltaMainRotor = MAX_MAIN_ROTOR_DELTA * elapsedTime;
		if (desMainRotorSpeed_RPM > MAX_MAIN_ROTOR_SPEED)
		{
			desMainRotorSpeed_RPM = MAX_MAIN_ROTOR_SPEED;
		}
		if (actMainRotorSpeed_RPM < desMainRotorSpeed_RPM)
		{
			actMainRotorSpeed_RPM += deltaMainRotor;
			if (actMainRotorSpeed_RPM > desMainRotorSpeed_RPM)
			{
				actMainRotorSpeed_RPM = desMainRotorSpeed_RPM;
			}
		}
		else if (actMainRotorSpeed_RPM > desMainRotorSpeed_RPM)
		{
			actMainRotorSpeed_RPM -= deltaMainRotor;
			if (actMainRotorSpeed_RPM < desMainRotorSpeed_RPM)
			{
				actMainRotorSpeed_RPM = desMainRotorSpeed_RPM;
			}
		}
	}
	
	public void updateTailRotorSpeed(double elapsedTime)
	{
		double deltaTailRotor = MAX_TAIL_ROTOR_DELTA * elapsedTime;
		if (desTailRotorSpeed_RPM > MAX_TAIL_ROTOR_SPEED)
		{
			desTailRotorSpeed_RPM = MAX_TAIL_ROTOR_SPEED;
		}
		if (actTailRotorSpeed_RPM < desTailRotorSpeed_RPM)
		{
			actTailRotorSpeed_RPM += deltaTailRotor;
			if (actTailRotorSpeed_RPM > desTailRotorSpeed_RPM)
			{
				actTailRotorSpeed_RPM = desTailRotorSpeed_RPM;
			}
		}
		else if (actTailRotorSpeed_RPM > desTailRotorSpeed_RPM)
		{
			actTailRotorSpeed_RPM -= deltaTailRotor;
			if (actTailRotorSpeed_RPM < desTailRotorSpeed_RPM)
			{
				actTailRotorSpeed_RPM = desTailRotorSpeed_RPM;
			}
		}
	}
	
	public void updateTiltLevel(double elapsedTime)
	{
		double deltaTailRotor = MAX_TAIL_ROTOR_DELTA * elapsedTime;
		if (desTilt_Degrees > MAX_TILT_MAGNITUDE)
		{
			desTilt_Degrees = MAX_TILT_MAGNITUDE;
		}
		if (desTilt_Degrees < -MAX_TILT_MAGNITUDE)
		{
			desTilt_Degrees = -MAX_TILT_MAGNITUDE;
		}
		double deltaTilt = MAX_TILT_DELTA * elapsedTime;
		if (actTilt_Degrees < desTilt_Degrees)
		{
			actTilt_Degrees += deltaTilt;
			if (actTilt_Degrees > desTilt_Degrees)
			{
				actTilt_Degrees = desTilt_Degrees;
			}
		}
		else if (actTilt_Degrees > desTilt_Degrees)
		{
			actTilt_Degrees -= deltaTilt;
			if (actTilt_Degrees < desTilt_Degrees)
			{
				actTilt_Degrees = desTilt_Degrees;
			}
		}
	}

	public boolean updateFuelRemaining(double elapsedTime)
	{
		boolean outOfGas = false;
		double rotorRevolutions = actMainRotorSpeed_RPM * elapsedTime;
		currentFuel_kg -= rotorRevolutions * FUEL_PER_REVOLUTION;
		if (currentFuel_kg < 0)
		{
			outOfGas = true;
		}
		return outOfGas;
		
	}
	
	public void updateCurrentHeading(double elapsedTime)
	{
		double rotationCalculator = actTailRotorSpeed_RPM;
		if (rotationCalculator < MIN_TAIL_ROTOR_SPEED)
		{
			rotationCalculator = MIN_TAIL_ROTOR_SPEED;
		}
		else if (rotationCalculator > MAX_TAIL_ROTOR_SPEED)
		{
			rotationCalculator = MAX_TAIL_ROTOR_SPEED;
		}
		double rotorSetting = rotationCalculator - STABLE_TAIL_ROTOR_SPEED;
		heading_Degrees += (rotorSetting * ROTATION_PER_TAIL_RPM) * elapsedTime;
	}
	
	public void fly(double elapsedTime)
	{
		boolean outOfGas = updateFuelRemaining(elapsedTime);
		if (outOfGas)
		{
			desMainRotorSpeed_RPM = 0.0;
			desTailRotorSpeed_RPM = 0.0;
			desTilt_Degrees = 0.0;
		}
		updateMainRotorSpeed(elapsedTime);
		updateTailRotorSpeed(elapsedTime);
		updateTiltLevel(elapsedTime);
		double cargoMass_kg = 0.0;
		StigChopper thisChopper = myWorld.getChopper(chopperID);
		if (thisChopper != null)
		{
			cargoMass_kg = ChopperAggregator.ITEM_WEIGHT * thisChopper.itemCount();
		}
		double totalMass_kg = cargoMass_kg + currentFuel_kg + ChopperAggregator.BASE_MASS;
		double downForce_N = totalMass_kg * EARTH_ACCELERATION;
		double actTilt_radians = actTilt_Degrees * Math.PI / 180.0;
		double liftForce_N = actMainRotorSpeed_RPM * THRUST_PER_RPM * Math.cos(actTilt_radians);
		// lateral force will only be used when off the ground (See below)
		double lateralForce_N = actMainRotorSpeed_RPM * THRUST_PER_RPM * Math.sin(actTilt_radians);
		double deltaForce_N = liftForce_N - downForce_N;
		if (deltaForce_N > 0.0) // We have enough force to ascend
		{
			// We know vertical force, we'll compute lateral forces next
			actAcceleration_ms2.m_z = deltaForce_N;
			if (takenOff == false)
			{
				takenOff = true;
			}
		}
		else
		{
			// TODO: Determine if we're sitting on the ground, for now, it's treated SIMPLE
			// This means that once you take off, you can go under the world
			if (takenOff == true)
			{
				actAcceleration_ms2.m_z = deltaForce_N;
			}
			else
			{
				actAcceleration_ms2.m_z = 0.0;
			}
		}
		if (takenOff) // Tail rotor comes into play
		{
			updateCurrentHeading(elapsedTime);
			// Now that we have our heading, we can compute the direction of our thrust
			double heading_radians = heading_Degrees * Math.PI / 180.0;
			actAcceleration_ms2.m_x = Math.cos(heading_radians);
			actAcceleration_ms2.m_y = Math.sin(heading_radians);
		}
		else
		{
			// For now, we're preventing skating -- chopper sliding along the ground
			actAcceleration_ms2.m_x = 0.0;
			actAcceleration_ms2.m_y = 0.0;
		}
		// now that accurate acceleration is computed, we can compute new velocity
		actVelocity_ms.m_x += (actAcceleration_ms2.m_x * elapsedTime);
		actVelocity_ms.m_y += (actAcceleration_ms2.m_y * elapsedTime);
		actVelocity_ms.m_z += (actAcceleration_ms2.m_z * elapsedTime);
		// Now that accurate velocity is computed, we can update position
		actPosition_m.m_x += (actVelocity_ms.m_x * elapsedTime);
		actPosition_m.m_y += (actVelocity_ms.m_y * elapsedTime);
		actPosition_m.m_z += (actVelocity_ms.m_z * elapsedTime);
		show();
	}
	
	public void show()
	{
		System.out.println("Main Rotor RPM: " + actMainRotorSpeed_RPM);
		System.out.println("Acceleration: " + actAcceleration_ms2.info());
	}
}

File: python/ChopperInfo.py
#!/usr/bin/python3

# (c) 2015-2024, A Beloussov, D. Lafuze

from panda3d.core import Vec3, Vec4

import math
import random


class ChopperInfo:
    def __init__(self, id, fuelCap, startPos, startHeading):
        self.TAG = "ChopperInfo"
        self.CI_DBG = 0x20000000
        self.THRUST_PER_RPM = 11.1111 # N (kg * m/s^2)
        self.MAX_MAIN_ROTOR_SPEED = 400.0 # RPM
        self.EARTH_ACCELERATION = 9.80665 # m/s^2
        self.MAX_TAIL_ROTOR_SPEED = 120.0 # RPM
        self.STABLE_TAIL_ROTOR_SPEED = 100.0 # RPM
        self.MIN_TAIL_ROTOR_SPEED = 80.0 # RPM
        self.MAX_TILT_MAGNITUDE = 10.0 # Degrees
        self.MAX_MAIN_ROTOR_DELTA = 60.0 # RPM per Second
        self.MAX_TAIL_ROTOR_DELTA = 30.0 # RPM per Second
        self.MAX_TILT_DELTA = 3.0 # Degrees per second
        self.FUEL_PER_REVOLUTION = 1.0 / 60.0 # Liters
        self.ROTATION_PER_TAIL_RPM = 3.0 # degrees per second

        self.chopperID = id
        self.mainRotorPosition_Degrees = 0.0 # This is needed to draw the rotor
        self.tailRotorPosition_Degrees = 0.0 # This is needed to draw the rotor
        
        self.actMainRotorSpeed_RPM = 0.0
        self.actTailRotorSpeed_RPM = 0.0
        self.actTilt_Degrees = 0.0
        self.desMainRotorSpeed_RPM = 0.0
        self.desTailRotorSpeed_RPM = 0.0
        self.desTilt_Degrees = 0.0
        self.remainingFuel_kg = fuelCap
        self.takenOff = False
        # In meters per second squared
        self.actAcceleration_ms2 = Vec3(0.0, 0.0, 0.0)
        # In meters per second
        self.actVelocity_ms = Vec3(0.0, 0.0, 0.0)
        # In meters
        self.actPosition_m = Vec4(startPos.getX(), startPos.getY(), startPos.getZ(),0.0)
        self.heading_Degrees = startHeading
        self.m_revs_sum = 0.0
        self.m_burnt_sum = 0.0
        self.m_time_sum = 0.0

    def getFuelRemaining(self):
        return self.remainingFuel_kg


    def getMainRotorPosition(self):
        return self.mainRotorPosition_Degrees
	
	
    def getTailRotorPosition(self):
        return self.tailRotorPosition_Degrees
	
        
    def getPosition(self):
        return self.actPosition_m
	
	
    def getHeading(self):
        return self.heading_Degrees
	
	
    def getTilt(self):
        return self.actTilt_Degrees
	
    def requestMainRotorSpeed(self, newSpeed):
        if (self.remainingFuel_kg > 0.0):
            self.desMainRotorSpeed_RPM = newSpeed
	
    def requestTailRotorSpeed(self, newSpeed):
        if (self.remainingFuel_kg > 0.0):
            self.desTailRotorSpeed_RPM = newSpeed
	
    def requestTiltLevel(self, newTilt):
        if (self.remainingFuel_kg > 0.0):
            self.desTilt_Degrees = newTilt
		
    def updateMainRotorSpeed(self, elapsedTime):
        deltaMainRotor = self.MAX_MAIN_ROTOR_DELTA * elapsedTime
        
        if (self.desMainRotorSpeed_RPM > self.MAX_MAIN_ROTOR_SPEED):
            self.desMainRotorSpeed_RPM = self.MAX_MAIN_ROTOR_SPEED
            
        if (self.actMainRotorSpeed_RPM < self.desMainRotorSpeed_RPM):
            self.actMainRotorSpeed_RPM += deltaMainRotor
            
            if (self.actMainRotorSpeed_RPM > self.desMainRotorSpeed_RPM):
                self.actMainRotorSpeed_RPM = self.desMainRotorSpeed_RPM
        
        elif (self.actMainRotorSpeed_RPM > self.desMainRotorSpeed_RPM):
        
            self.actMainRotorSpeed_RPM -= deltaMainRotor
            if (self.actMainRotorSpeed_RPM < self.desMainRotorSpeed_RPM):
                self.actMainRotorSpeed_RPM = self.desMainRotorSpeed_RPM
            
        
        # 1 RPM = 6 degrees per second
        self.mainRotorPosition_Degrees += self.actMainRotorSpeed_RPM * elapsedTime * 60.0
        while (self.mainRotorPosition_Degrees >= 360.0):
            self.mainRotorPosition_Degrees -= 360.0 # Just for drawing
	
    def updateTailRotorSpeed(self, elapsedTime):
        
        deltaTailRotor = self.MAX_TAIL_ROTOR_DELTA * elapsedTime
        if (self.desTailRotorSpeed_RPM > self.MAX_TAIL_ROTOR_SPEED):
        
            self.desTailRotorSpeed_RPM = self.MAX_TAIL_ROTOR_SPEED
        
        if (self.actTailRotorSpeed_RPM < self.desTailRotorSpeed_RPM):
        
            self.actTailRotorSpeed_RPM += deltaTailRotor
            if (self.actTailRotorSpeed_RPM > self.desTailRotorSpeed_RPM):
            
                self.actTailRotorSpeed_RPM = self.desTailRotorSpeed_RPM
            
        
        elif (self.actTailRotorSpeed_RPM > self.desTailRotorSpeed_RPM):
        
            self.actTailRotorSpeed_RPM -= deltaTailRotor
            if (self.actTailRotorSpeed_RPM < self.desTailRotorSpeed_RPM):
            
                self.actTailRotorSpeed_RPM = self.desTailRotorSpeed_RPM
        
        # 1 RPM = 6 degrees per second
        self.tailRotorPosition_Degrees += self.actTailRotorSpeed_RPM * elapsedTime * 60.0
        while (self.tailRotorPosition_Degrees >= 360.0):
        
            self.tailRotorPosition_Degrees -= 360.0 # Just for drawing
		
	
	
    def updateTiltLevel(self, elapsedTime):
        
        deltaTailRotor = self.MAX_TAIL_ROTOR_DELTA * elapsedTime
        if (self.desTilt_Degrees > self.MAX_TILT_MAGNITUDE):
        
            self.desTilt_Degrees = self.MAX_TILT_MAGNITUDE
        
        if (self.desTilt_Degrees < -self.MAX_TILT_MAGNITUDE):
        
            self.desTilt_Degrees = -self.MAX_TILT_MAGNITUDE
        
        deltaTilt = self.MAX_TILT_DELTA * elapsedTime
        if (self.actTilt_Degrees < self.desTilt_Degrees):
        
            self.actTilt_Degrees += deltaTilt
            if (self.actTilt_Degrees > self.desTilt_Degrees):
                self.actTilt_Degrees = self.desTilt_Degrees
            
        elif (self.actTilt_Degrees > self.desTilt_Degrees):
        
            self.actTilt_Degrees -= deltaTilt
            if (self.actTilt_Degrees < self.desTilt_Degrees):
            
                self.actTilt_Degrees = self.desTilt_Degrees
			
		
	
    def updateFuelRemaining(self, elapsedTime):
        
        outOfGas = False
        rotorRevolutions = self.actMainRotorSpeed_RPM / 60.0 * elapsedTime
        fuelBurned = rotorRevolutions * self.FUEL_PER_REVOLUTION
        self.remainingFuel_kg -= fuelBurned
        self.m_revs_sum += rotorRevolutions
        self.m_burnt_sum += fuelBurned
        self.m_time_sum += elapsedTime
        if (self.remainingFuel_kg < 0):
        
            base.dbg(self.TAG,"Out of Gas!",self.CI_DBG)
            self.remainingFuel_kg = 0.0
            outOfGas = True
        
        return outOfGas
		
	
	
    def updateCurrentHeading(self, elapsedTime):
        
        rotationCalculator = self.actTailRotorSpeed_RPM
        if (rotationCalculator < self.MIN_TAIL_ROTOR_SPEED):
        
            rotationCalculator = self.MIN_TAIL_ROTOR_SPEED
        
        elif (rotationCalculator > self.MAX_TAIL_ROTOR_SPEED):
        
            rotationCalculator = self.MAX_TAIL_ROTOR_SPEED
        
        rotorSetting = rotationCalculator - self.STABLE_TAIL_ROTOR_SPEED
        self.heading_Degrees += (rotorSetting * self.ROTATION_PER_TAIL_RPM) * elapsedTime
        while (self.heading_Degrees > 360.0):
        
            self.heading_Degrees -= 360.0
        
        while (self.heading_Degrees < 0.0):
        
            self.heading_Degrees += 360.0
		
	
	
    def fly(self, currentTime,  elapsedTime):
        outOfGas = self.updateFuelRemaining(elapsedTime)
        if (outOfGas):
            self.desMainRotorSpeed_RPM *= 0.99
            self.desTailRotorSpeed_RPM *= 0.99
            self.desTilt_Degrees += -1.5 + 2.0 * random.randint(0,100) * 0.01
        
        self.updateMainRotorSpeed(elapsedTime)
        self.updateTailRotorSpeed(elapsedTime)
        self.updateTiltLevel(elapsedTime)
        cargoMass_kg = 0.0
        thisChopper = base.getChopper(self.chopperID)
        if (thisChopper is not None):
            cargoMass_kg = base.ITEM_WEIGHT * thisChopper.itemCount()
        
        totalMass_kg = cargoMass_kg + self.remainingFuel_kg + base.CHOPPER_BASE_MASS
        downForce_N = totalMass_kg * self.EARTH_ACCELERATION # F = mA
        actTilt_radians = math.radians(self.actTilt_Degrees)
        liftForce_N = self.actMainRotorSpeed_RPM * self.THRUST_PER_RPM * math.cos(actTilt_radians)
        # lateral force will only be used when off the ground (See below)
        lateralForce_N = self.actMainRotorSpeed_RPM * self.THRUST_PER_RPM * math.sin(actTilt_radians)
        lateralAcceleration = lateralForce_N / totalMass_kg
        deltaForce_N = liftForce_N - downForce_N
        if (deltaForce_N > 0.0): # We have enough force to ascend
            # We know vertical force, we'll compute lateral forces next
            self.actAcceleration_ms2.setZ(deltaForce_N / totalMass_kg)
            if (self.takenOff == False):
                base.dbg(self.TAG,f"Chopper {self.chopperID} has lifted off!",self.CI_DBG)
                self.takenOff = True
        else: 
            # Simple landing check when close to zero
            if (self.actPosition_m.getZ() < 0.25):
                lateralMagnitude = self.actVelocity_ms.getXy().length()
                base.dbg(self.TAG, "Chopper " + str(self.chopperID) + " Landing check lateral velocity: " + str(lateralMagnitude) + ", vert Velocity: " + str(self.actVelocity_ms.getZ()), self.CI_DBG)
                if (lateralMagnitude < 0.25 and (self.actVelocity_ms.getZ() > (-2.0) and self.actVelocity_ms.getZ() < 0)):
                
                    if (self.takenOff == True):
                        base.dbg(self.TAG,f"Chopper {self.chopperID} has landed!",self.CI_DBG)
                    self.takenOff = False
                
            
            if (self.takenOff == True):
                self.actAcceleration_ms2.setZ( deltaForce_N / totalMass_kg )
            else: 
                self.actAcceleration_ms2.setZ(0.0)
                self.actVelocity_ms.setZ(0.0)
                self.actPosition_m.setZ(0.0)
            
        if (self.takenOff): # Tail rotor comes into play
            self.updateCurrentHeading(elapsedTime)
            # Now that we have our heading, we can compute the direction of our thrust
            heading_radians = math.radians(self.heading_Degrees)
            self.actAcceleration_ms2.setX(lateralAcceleration * math.sin(heading_radians))
            self.actAcceleration_ms2.setY(lateralAcceleration * math.cos(heading_radians))
        else: 
            # For now, we're preventing skating -- chopper sliding along the ground
            self.actAcceleration_ms2.setX(0.0)
            self.actAcceleration_ms2.setY(0.0)
            self.actVelocity_ms.setX(0.0)
            self.actVelocity_ms.setY(0.0)
        
        # now that accurate acceleration is computed, we can compute new velocity
        self.actVelocity_ms.setX((self.actVelocity_ms.getX() + self.actAcceleration_ms2.getX() * elapsedTime))
        self.actVelocity_ms.setY((self.actVelocity_ms.getY() + self.actAcceleration_ms2.getY() * elapsedTime))
        self.actVelocity_ms.setZ((self.actVelocity_ms.getZ() + self.actAcceleration_ms2.getZ() * elapsedTime))
        # Now that accurate velocity is computed, we can update position
        self.actPosition_m.addX((self.actVelocity_ms.getX() * elapsedTime))
        self.actPosition_m.addY((self.actVelocity_ms.getY() * elapsedTime))
        self.actPosition_m.addZ((self.actVelocity_ms.getZ() * elapsedTime))
        self.actPosition_m.setW(currentTime)
	
    def onGround(self,):
        return not self.takenOff
	
    def show(self, curTime):
        base.dbg(self.TAG,f"Heading: {self.heading_Degrees} deg, desired rotor speed: {self.desMainRotorSpeed_RPM}",self.CI_DBG)
        #World.dbg(self.TAG,"World Time: " + curTime + ", Acceleration: " + self.actAcceleration_ms2.info(),self.CI_DBG)
        #World.dbg(self.TAG,"Actual Heading: " + self.heading_Degrees + " Degrees, Velocity: " + self.actVelocity_ms.info(),self.CI_DBG)
        #World.dbg(self.TAG,"Actual Tilt: " + self.actTilt_Degrees + " Degrees, Position: " + self.actPosition_m.info(),self.CI_DBG)
	


	
	